webapp-django/crashstats/home/static/home/js/home.js

```diff
@@ -0,0 +1,386 @@
+            }
+        }
+
+        // Now transform that data into a format that correspond to what
+        // the graph library will be expecting.
+        var superSearchData = transformSuperSearchData(dataStruct);
+
+        dataStore.superSearch = superSearchData;
+        onDataComplete();
+    }
+
+    function transformSuperSearchData(dataStruct) {
+        var superSearchData = [];
+        for (var i = 0, ln = versions.length; i < ln; i++) {
+            superSearchData[i] = [];
+        }
+
+        for (var version in dataStruct) {
+            var versionIndex = versions.indexOf(version);
+
+            for (var date in dataStruct[version]) {
+                superSearchData[versionIndex].push({
+                    date: new Date(date),
+                    count: dataStruct[version][date],
+                });
+            }
+        }
+
+        return superSearchData;
+    }
+
+    /**
+     * Format data received from ADI to make it easier to exploit.
+     */
+    function onAdiData(data) {
+        var adiData = {};
+        for (var i = 0, ln = data.hits.length; i < ln; i++) {
+            var dayCount = data.hits[i];
+            if (!adiData[dayCount.date]) {
+                adiData[dayCount.date] = {};
+            }
+
+            if (versions.indexOf(dayCount.version) > -1) {
+                adiData[dayCount.date][dayCount.version] = {
+                    count: dayCount.adi_count,
+                    build: dayCount.build_type,
+                };
+            }
+
+            // Handle beta versions.
+            for (var k = betaVersions.length - 1; k >= 0; k--) {
+                if (dayCount.version.indexOf(betaVersions[k]) > -1) {
+                    if (!adiData[dayCount.date][betaVersions[k]]) {
+                        adiData[dayCount.date][betaVersions[k]] = {
+                            count: 0,
+                            build: dayCount.build_type,
+                        };
+                    }
+
+                    adiData[dayCount.date][betaVersions[k]].count += dayCount.adi_count;
+                }
+            }
+        }
+        dataStore.adi = adiData;
+        onDataComplete();
+    }
+
+    /**
+     * Simply pass the data about build types.
+     */
+    function onBuildTypesData(data) {
+        dataStore.builds = data.hits;
+        onDataComplete();
+    }
+
+    /**
+     * Show an error if any of the required services failed.
+     */
+    function onQueryError(jqXHR, textStatus, errorThrown) {
+        $('.loading').remove();
+
+        container
+            .empty()
+            .append('There was an error processing the request: Status: ' + textStatus + ' Error: ' + errorThrown);
+    }
+
+    /**
+     * Called every time a service has stored some data. When all expected
+     * data is received, proceed with mixing that data and showing the graph.
+     */
+    function onDataComplete() {
```

Comment:
This needs to be entirely replaced with promises and deferreds. jQuery's built-in is quite nice. 
You're currently relying on `dataStore.superSearch` to exist. Which is kinda weird. Unlike Python, an array is always true. Even if there's nothing in there. Same for a dictionary. So if `dataStore.superSearch` is `[]` or `{}` this'll be true. 

I was going to say something stupid like "Check the documentation on $.when". But to back up my feedback, I put together a little test to make sure I understand this myself. 

http://codepen.io/peterbe/pen/wMLLwe?editors=0010

That way you get a promise call back once all the **post-processing** has been done **after** the AJAX has finished, for each AJAX query.